src/content/popup/KanjiEntryTopPart.tsx

```diff
@@ -0,0 +1,291 @@
+      <div>
+        <Readings entry={entry} />
+
+        {/* -- -- Meta */}
+        {entry.misc.meta && <Meta meta={entry.misc.meta} />}
+
+        {/* -- -- Meanings */}
+        <div className="meanings" lang={entry.m_lang}>
+          {entry.m.join(', ')}
+        </div>
+
+        {/* -- -- Misc info */}
+        <MiscRow entry={entry} />
+
+        {/* -- -- Kanji components */}
+        {(typeof options.showKanjiComponents === 'undefined' ||
+          options.showKanjiComponents) && <KanjiComponents entry={entry} />}
+      </div>
+    </div>
+  );
+}
+
+function Readings({ entry }: { entry: KanjiResult }) {
+  type Reading = string | { kanjiReading: string; okuriganaReading: string };
+  const readings: Array<Reading> = [];
+
+  if (entry.r.on && entry.r.on.length) {
+    readings.push(...entry.r.on);
+  }
+
+  // Kun readings sometimes have a . in them separating the initial part that
+  // represents the kanji, from the okurigana.
+  //
+  // e.g. あた.える
+  //
+  // We want to take the bit after the '.' and separate it, so that it can be
+  // wrapped in a span with an appropriate class below.
+  for (const reading of entry.r.kun || []) {
+    const highlightIndex = reading.indexOf('.');
+    if (highlightIndex === -1) {
+      readings.push(reading);
+    } else {
+      readings.push({
+        kanjiReading: reading.substring(0, highlightIndex),
+        okuriganaReading: reading.substring(highlightIndex + 1),
+      });
+    }
+  }
+
+  const { t, langTag } = useLocale();
+
+  return (
+    <div className="readings" lang="ja">
+      {readings
+        .map<ReactNode>((reading) => {
+          if (typeof reading === 'string') {
+            return reading;
+          } else {
+            return (
+              <>
+                {reading.kanjiReading}
+                <span class="okurigana">{reading.okuriganaReading}</span>
+              </>
+            );
+          }
+        })
+        .reduce((prev, curr) => [prev, '、', curr])}
+
+      {entry.r.na && entry.r.na.length && (
+        <>
+          <br />
+          <span className="nanorilabel" lang={langTag}>
+            {t('content_kanji_nanori_label')}
+          </span>
+          {` ${entry.r.na.join('、')}`}
+        </>
+      )}
+    </div>
+  );
+}
+
+function Meta({ meta }: { meta?: string[] }) {
+  if (!meta) {
+    return null;
+  }
+
+  const { t, langTag } = useLocale();
+  return (
+    <div className="meta">
+      {meta.map((tag) => (
+        <span className="tag" lang={langTag}>
+          {t(`content_kanji_meta_${tag.replace(' ', '_')}`)}
+        </span>
+      ))}
+    </div>
+  );
+}
+
+function LegacyIcon({
+  iconSvg,
+  opacity,
+}: {
+  iconSvg: SVGElement;
+  opacity: number;
+}) {
+  const ref = useRef<HTMLDivElement>(null);
+  useLayoutEffect(() => {
+    iconSvg.classList.add('svgicon');
+    iconSvg.style.opacity = `${opacity}`;
+    ref.current?.replaceWith(iconSvg);
+  });
+
+  return <div ref={ref}></div>;
+}
```

Comment:
This is a temporary hack I have no intention of actually keeping. I just did it to get the icons rendering onscreen again with the new React setup, without having to change too much code at once.

However, it does raise the question of how you'd like to handle the icons in `icons.ts`. Should I just convert each of those to a React component that returns an `<svg>`?

Reply:
> However, it does raise the question of how you'd like to handle the icons in `icons.ts`. Should I just convert each of those functions to a React component that returns an `<svg>`?

Yes, I think converting those to individual React components makes the most sense.